### ISSUE TITLE
Test fails when -Xno-call-assertions is omitted

### ISSUE THREAD

**Issue by shmuelr** [Kotlin/kotlinx.coroutines#2209]
For context, I'm compiling Kotlin Coroutines from source instead of the prebuilt artifacts (due to monorepo development) and hit this issue when migrating our codebase to the latest Coroutine version. I created a reproducible test that only passes when `-Xno-call-assertions` is used:

```
    @Test
    fun foo() = runTest {
        assertEquals(listOf(999), testFlow().toList())
    }

    private fun testFlow(): Flow<Int> {
        return combine(
            flowOne(),
            flowTwo().map { it.getOrNull() }
        ) { first: Int, second: Int? ->
            second ?: return@combine 999
            first
        }
    }

    private fun flowOne(): Flow<Int> {
        return flow {
            emit(1)
        }
    }

    private fun flowTwo(): Flow<Result<Int>> {
        return flow {
            emit(Result.failure<Int>(RuntimeException()))
        }
    }
```

This test should pass because `second` will be null due to the `getOrNull()` call for the `Result.failure()`. If `-Xno-call-assertions` is omitted this test fails with 

```
transform(value) must not be null
java.lang.NullPointerException: transform(value) must not be null
	at kotlinx.coroutines.flow.FlowInvariantsTest$testFlow$$inlined$map$1$2.emit(Collect.kt:135)
	at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt:15)
	at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt)
	at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:73)
	at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:55)
	at kotlinx.coroutines.flow.FlowInvariantsTest$flowTwo$1.invokeSuspend(FlowInvariantsTest.kt:285)
```

It seems the the code in [Collect.kt](https://github.com/Kotlin/kotlinx.coroutines/blob/583ec6e862fb70c7fc0232af948a028bab3b20de/kotlinx-coroutines-core/common/src/flow/terminal/Collect.kt#L73) is expecting to be non null, but the compiler is allowing nulls there if call assertions are enabled. 

Sample code:
https://github.com/shmuelr/kotlinx.coroutines/commit/486c6b447a4b16e1ef002ce329c6104b93283359

I'm running this on JDK 11

Thank you!

**Comment by kevin1e100** [comment]
We believe this issue is similarly avoided with `-XXLanguage:-NewInference` instead of `-Xno-call-assertions`, suggesting there may be a compiler issue causing this?

**Comment by qwwdfsad** [comment]
We've investigated that the source of this behaviour is compiler bug related to using `Result` type 
As a workaround, it's enough to wrap `Result` into non-inline type

ETA of fix is Kotlin 1.4.20.

**Comment by kevin1e100** [comment]
Thanks for the quick investigation.  Is there bug for the underlying issue that we can follow?  I'd like to understand how much this might affect uses of `Result` outside of coroutines, or even uses of other inline classes.

**Comment by qwwdfsad** [comment]
We had an in-depth investigation internally but unfortunately didn't document it properly in a public space.
https://youtrack.jetbrains.com/issue/KT-41429 is created to keep track of the progress, but no detailed explanation is provided, unfortunately.

**Comment by davidjwiner** [comment]
This issue is unfortunately blocking all Google teams from updating to Kotlin 1.4 and I imagine any others who are trying to build coroutines from source. Is there any way we could prioritize this for the 1.4.1 release?

**Comment by pyos** [comment]
This seems to be a problem with `BuilderInference`, not suspend functions. The following code also fails:

```
@file:OptIn(kotlin.experimental.ExperimentalTypeInference::class)

interface Flow<out T> {
    fun collect(collector: FlowCollector<T>)
}

interface FlowCollector<in T> {
    fun emit(value: T)
}

fun <T> flow(block: FlowCollector<T>.() -> Unit): Flow<T> =
    object : Flow<T> {
        override fun collect(collector: FlowCollector<T>) = collector.block()
    }

fun <T> Flow<T>.collect(action: (value: T) -> Unit): Unit =
    collect(object : FlowCollector<T> {
        override fun emit(value: T) = action(value)
    })

fun <T, R> Flow<T>.transform(@BuilderInference transform: FlowCollector<R>.(T) -> Unit): Flow<R> =
    flow { collect { transform(it) } }

fun <T, R> Flow<T>.map(transform: (T) -> R): Flow<R> =
    transform { emit(transform(it)) }

fun main() {
    flow<Int> { emit(1) }.map { null }.collect { println(it) }
}
```

The null check is already present in `map`.

**Comment by zarechenskiy** [comment]
Thanks for the example, it turns to be a problem in FE. We're going to ship a fix in 1.4.10: https://youtrack.jetbrains.com/issue/KT-41470

**Comment by shmuelr** [comment]
I can confirm this is fixed with the new [Kotlin 1.4.10 release](https://github.com/JetBrains/kotlin/releases/tag/v1.4.10) 

Thanks for the quick fix!
